go-controller/pkg/ovn/external_ids_syncer/port_group/port_group_sync.go

```diff
@@ -0,0 +1,230 @@
+		// create updated port group
+		ops, err = libovsdbops.CreateOrUpdatePortGroupsOps(syncer.nbClient, ops, portGroupInfo.newPG)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get update port group ops for port group %s: %v", oldName, err)
+		}
+		// delete old port group
+		ops, err = libovsdbops.DeletePortGroupsOps(syncer.nbClient, ops, portGroupInfo.oldPG.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get update port group ops for port group %s: %v", oldName, err)
+		}
+		oldHash := "@" + portGroupInfo.oldPG.Name
+		newHash := "@" + portGroupInfo.newPG.Name
+
+		for _, acl := range portGroupInfo.acls {
+			if _, ok := aclsToUpdate[acl.UUID]; !ok {
+				aclsToUpdate[acl.UUID] = acl
+			}
+			aclsToUpdate[acl.UUID].Match = strings.ReplaceAll(aclsToUpdate[acl.UUID].Match, oldHash, newHash)
+		}
+	}
+
+	for _, acl := range aclsToUpdate {
+		ops, err = libovsdbops.UpdateACLsOps(syncer.nbClient, ops, acl)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get update acl ops: %v", err)
+		}
+	}
+
+	return
+}
+
+// getPortGroupUpdateInfo adds db ops to update port group and objects that reference it
+func (syncer *PortGroupSyncer) getPortGroupUpdateInfo(pg *nbdb.PortGroup) (*updatePortGroupInfo, error) {
+	pgName := pg.ExternalIDs["name"]
+	networkExternalID := pg.ExternalIDs[types.NetworkExternalID]
+	if pgName == "" {
+		return nil, fmt.Errorf("port group doesn't have expecter ExternalID[\"name\"]")
+	}
+
+	acls, dbIDs, err := syncer.getReferencingObjsAndNewDbIDs(pg.Name, pgName, networkExternalID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get new dbIDs for port group %s: %v", pg.ExternalIDs["name"], err)
+	}
+
+	// since we need to update portGroup.Name, which is an index and not listed in getAllUpdatableFields,
+	// we copy existing portGroup, update the required fields, and replace (delete and create) existing port group with the updated
+	newPG := pg.DeepCopy()
+	// reset UUID
+	newPG.UUID = ""
+	// update port group Name and ExternalIDs exactly the same way as in libovsdbops.BuildPortGroup
+	newPG.Name = libovsdbops.GetPortGroupName(dbIDs)
+	newPG.ExternalIDs = dbIDs.GetExternalIDs()
+	return &updatePortGroupInfo{acls, pg, newPG}, nil
+}
+
+// SyncPortGroups must be run after ACLs sync, since it uses new ACL.ExternalIDs
+func (syncer *PortGroupSyncer) SyncPortGroups() error {
+	// stale port groups don't have controller ID
+	portGroupList, err := libovsdbops.FindPortGroupsWithPredicate(syncer.nbClient, libovsdbops.GetNoOwnerPredicate[*nbdb.PortGroup]())
+	if err != nil {
+		return fmt.Errorf("failed to find stale port groups: %v", err)
+	}
+
+	updatedCount := 0
+	defer func() {
+		klog.Infof("SyncPortGroups handled %d of %d stale port groups", updatedCount, len(portGroupList))
+	}()
+	i := 0
+	for i < len(portGroupList) {
+		pgUpdateInfos := []*updatePortGroupInfo{}
+		for j := 0; (j < syncer.txnBatchSize || syncer.txnBatchSize == 0) && i < len(portGroupList); i, j = i+1, j+1 {
```

Comment:
This looks weird. Why can't we use the Batch function itself?

Reply:
I had to dig the history for it :D So when we merged first external_ids_syncer for address set, batching didn't exist. Then for acls it already was used. I copy-pasted something form the address sets here, that is how we ended up with this code.
I will add an extra commit first to update address sets syncer to use batching, and then update this one in the existing commit